### DISCUSSION TITLE
Optimizing Galaxy for Teaching: what could be improved?

### DISCUSSION THREAD

**Discussion by shiltemann** [galaxyproject/training-material#2198] [Instructors]
As instructors, we have a unique perspective by frequently watching people use Galaxy for analysis and/or for the first time. Did you observe them struggling with specific tasks? Have ideas for how to streamline how people interact with Galaxy? Missing features?  Any other ideas for how Galaxy can be improved for end-users? Please let us know here!

We will communicate these training priorities to the Galaxy Development team, see this [project board](https://github.com/orgs/galaxyproject/projects/7#card-48764289)

**Comment by shiltemann** [comment]
When users look at their multi-history view, they often ask if they can see the creation date of the history. I have been telling them to name their histories with the date, but perhaps the overview can show creation date and possibly a "last updated on" date at the top?

![image](https://user-images.githubusercontent.com/2563865/101890032-72195100-3ba0-11eb-968c-949cf30e974c.png)

**Reply by hexylena** [reply]
Yes! This should be a super easy paper cut. It looks like there's already an open issue from @fubar2 https://github.com/galaxyproject/galaxy/issues/10455

**Comment by yvanlebras** [comment]
Maybe THE major point I often heard from users / trainees, is related to the fact their is too much tools and this is hard to find the right tool... I konw that this is an old issue, and there was a lot of improvements, but it is still a real concern... Additionnaly to that, the search functionality is often not very usefull as let too much unrelated tools on the filtered list and/or the one searched is not there for obvious reason... Is there any manner to propose a maybe more complex / detailled  mutlicriteria tool serach form ? Or others ? Thanks Saskia to create this discussion, and sorry if it is not supposed to be used directly as I do ;)

**Reply by shiltemann** [reply]
@yvanlebras yes, this is EXACTLY what this discussion is for, thanks! We will communicate these obseservations from trainers to the Galaxy UI/UX team regularly so that they are aware of some of the pain points experienced by users.

I agree, especially for new users, the tool form feels very overwhelming, it would be great if they could have more control over this. Starring/Favoriting tools is definitely a step in the right directlon, but perhaps this can be expanded to support a custom user-defined view of the tool panel, where users can add tools from the full list of available tools orso, create their own section names, etc, to build up their own tool panel?

**Reply by abretaud** [reply]
Maybe it would be possible to build smaller tool panels based on the tool list of each tutorial?

**Reply by hexylena** [reply]
We've definitely discussed that @abretaud on the GTN side, it would be super cool. What's missing is that Galaxy knows which tutorial you're doing. It gets closer with the GTN-in-Galaxy, but, the connection is still missing and I'm not sure of a good way to integrate that :(

With the GTN in Galaxy webhook we could show a list of the tools on the left like a mini tool panel/ but still not super nice....

**Comment by shiltemann** [comment]
Watching some of the videos for Smorgasbord, I noticed a lot of instructors spend quite some time renaming datasets after import from URL. It would be nice if the upload form included an option to strip the full URL, and keeping only the filename. (and perhaps capturing the full URL as an annotation orso to save the info about its origin)

**Reply by hexylena** [reply]
xref https://github.com/galaxyproject/galaxy/issues/4005

**Comment by nomadscientist** [comment]
I'd like the option to rename a dataset before hitting RUN on a tool. That actually drags a surprisingly annoying amount of time in running a tutorial, because the tool finally finishes and then BAM you have to run a data change again. Also, I found a glitch with Manipulate AnnData wherein it runs fine but you can't rename it, so might avoid those issues in the future. All in all, might help keep people more organised and makes generating labels part of the tool rather than an overly zealous addition

**Comment by nomadscientist** [comment]
Getting different colour options in workflows would help for showing different sections and overall using the nice workflow editor in teaching

**Reply by hexylena** [reply]
xref https://github.com/galaxyproject/galaxy/issues/1726 https://github.com/galaxyproject/galaxy/issues/7090

**Comment by nomadscientist** [comment]
Jupyter in Galaxy - it is always a bit weird with interactive environments as they look perpetually on in the history, even if you close them. Trying to share a notebook, for example, will fail the second you stop your own environment, so sharability isn't great unless you leave it open infinitely.
Basically I've been building a tutorial and lost my notebook twice frogging around with 'stop' in the Active InteractiveTools bit and would like to make that easier in the future.

**Reply by bgruening** [reply]
@nomadscientist the intention is that you export your notebook to your history. And share the notebook file, the ipynb file.

**Reply by nomadscientist** [reply]
Ahhh ok. Fair enough. I think it's easy to get lax with Galaxy being so good at saving everything, that having to remember to save frequently AND export in the Jupyter environment can get you, but that does work and makes sense. 
Follow-up 1: Is it possible to 'label' a JupyterLab environment? [I may have stopped the wrong one at one point and am forever scarred]
Follow-up 2: Is it possible to get the training tutorial button IN the Jupiter environment as well? So they can continue on a Galaxy tutorial whilst in the notebook?

**Reply by hexylena** [reply]
1. We should list which history the container is running in / which HID. that would help.
2. This is more difficult. I think we should fix jupyter not running in the center panel, and bring back the path-based routing that allowed that to work in the original GIEs before it was dropped with GxITs :( then we wouldn't need to customise each container with a GTN overlay.

**Reply by nomadscientist** [reply]
1. Yes!
2. Yay!
3. Importing multiple datasets into an environment doesn't work. I know there is the option of getting them from within the environment, but it would be nice to be able to pull, say, a dataset and a notebook in at the same time.

**Reply by nomadscientist** [reply]
@bgruening I can confirm that this does not work within Galaxy. So if you push your notebook to your galaxy history, close the visualisation and/or share your history with someone else, they cannot open that notebook in Galaxy JupyterLab (either by calling it as a notebook or by calling it a dataset). Either way it glitches out. The only way to do it (as far as I can tell) is to download the notebook onto your computer, then upload it within the JupyterLab.

**Comment by nomadscientist** [comment]
Retroactively adding tags would be amazing - "Oh no, my history is a mess, I have two different paths running and I keep confusing them, how will I fix this?" --> goes to step 2, where I added a normalisation parameter differently for one set of analysis to compare with another, and adds a tag, and magically all of the following datasets that originated from that one suddenly have a tag. Glorious.

**Reply by anuprulez** [reply]
You can try to use the script: https://github.com/galaxyproject/galaxy/blob/dev/scripts/apply_tags.py to retroactively add tags to datasets in Galaxy history.

**Reply by nomadscientist** [reply]
Is that a tool? Is there a tool 'apply tags'? That would be amazing

**Reply by hexylena** [reply]
It's not a tool, it's a script within the galaxy codebase.  So it's a bit less easy to run for non-techie users, we should really fix this in the UI side of galaxy :(

The tool "apply tags" in galaxy does something confusingly different! hah

**Comment by nomadscientist** [comment]
'Hands-On' that can collapse down (like question/answers, but auto-set to be expanded). Just, you tend to scroll a lot with tutorials, so that might be handy

**Reply by nomadscientist** [reply]
Like 'yes, did that, tick' <--- or even a tick box omg

**Reply by hexylena** [reply]
Oh I like the tick box idea! It's really nice.

**Comment by nomadscientist** [comment]
It would be nice if ORCID somehow picked up the GTN tutorials/slide decks etc. Currently it doesn't, so I need to make a Bibtext thing and import it and it's a whole debacle.

**Reply by hexylena** [reply]
There's the automatic bibtex thing at the bottom, or does that not help?

![image](https://user-images.githubusercontent.com/458683/131315010-abf53fd3-96ee-47fa-a56e-04a2449e0397.png)

Automatically submitting data to ORCID's API however requires a "member" api key, and their membership is not cheap :( https://info.orcid.org/about-membership/ so looks like that's a no-go

**Reply by nomadscientist** [reply]
That is RUBBISH. Bad orchid

**Comment by nomadscientist** [comment]
<img width="199" alt="Screenshot 2022-01-17 at 17 45 11" src="https://user-images.githubusercontent.com/44605769/149817098-459e3b7e-9f72-4372-9426-f886e2dcf30e.png">
It would be awesome if tabular datasets (and similar) had both 'lines' AND 'columns'

**Reply by hexylena** [reply]
Filed as https://github.com/galaxyproject/galaxy/issues/13258

**Comment by nomadscientist** [comment]
The EYE button should look DIFFERENT depending on whether it starts DOWNLOADING a dataset vs SHOWS you the dataset
This would stop so many unnecessary downloads (by me, let alone my students)

**Reply by hexylena** [reply]
It really really should only ever *show* the dataset, is there a specific datatype that is resulting in a download?

**Reply by nomadscientist** [reply]
AnnData - https://usegalaxy.eu/u/wendi.bacon/h/imported-spanscriptomics-continuacin
scroll down to the 'Manipulate AnnData' step in the history and that auto-downloads if you click the eye

**Reply by hexylena** [reply]
Ahh it's an `h5ad` datatype. Not surprising that results in a download. I'll file a galaxy issue. https://github.com/galaxyproject/galaxy/issues/13492

**Reply by nomadscientist** [reply]
Oh also rdata - see 5:55s here: https://www.youtube.com/watch?v=nfAUXP3W-jM

**Comment by nomadscientist** [comment]
Workflow ranking!
For doing stuff like hackathons/project-based-learning/'intermediate/advanced' teaching, it could be awesome to have your listed workflows (obviously with stunning metadata, like 'cancer' embedded in them, of course...) and then participating can pick which workflows to try out on their data.
BUT lots of things break/fail/tools struggle
So maybe a 'voting' system on workflows ('yes this worked great!' / 'no this didn't work') so participants can more easily see what to use or not.

**Reply by nomadscientist** [reply]
Incidentally could help with other stuff, example - could embed website view of searching workflows for 'single-cell' on the single-cell training page
